editors/code/src/command/workspace-folder-formatting.ts

```diff
@@ -0,0 +1,228 @@
+import * as cp from "child_process";
+import * as vscode from "vscode";
+
+import { Cmd, Ctx } from "../context";
+import * as output from "../output";
+
+/**
+ * Format a workspace folder
+ *
+ * # Workspace folder selection
+ *
+ * - If 0 workspace folders are open, errors
+ * - If 1 workspace folder is open, automatically uses it
+ * - If >1 workspace folders are open, asks the user to choose
+ *
+ * # Tab closing
+ *
+ * Because we use the embedded Air CLI to perform the formatting, we force all
+ * relevant tabs to be saved and closed before proceeding. This has two main
+ * benefits:
+ *
+ * - The Air CLI can't know about any in-memory changes that may have occurred
+ *   in "dirty" editors. Forcing them to save and close causes any changes
+ *   to be written to disk first, syncing VS Code with the CLI.
+ *
+ * - The LSP server part of Air needs to perfectly track changes for any open
+ *   files. Closing the tabs first switches the server's "source of truth" from
+ *   the client version of the file to the on-disk version of the file, which
+ *   should ensure the LSP server won't have any chance of getting out of sync.
+ *   This may be overkill, since the LSP server seems to handle git branch
+ *   changes well, but it's still not a bad idea.
+ *
+ * # Forward propagation of settings
+ *
+ * If no `air.toml` is tied to a document, then typically in the LSP document
+ * settings are forward propagated from the client to the server, and those are
+ * used during formatting when that document is open. If an `air.toml` does
+ * exist, those settings override document settings.
+ *
+ * Because we go through the Air CLI here, forward propagated document settings
+ * are ignored, so even if you don't have an `air.toml`, you will get Air's
+ * default indentation settings.
+ *
+ * We can't simply require an `air.toml` at the workspace folder root to avoid
+ * confusion here, because you may instead have an `air.toml` at
+ * `{root}/R/air.toml` that would handle any R files in your project. Instead,
+ * we hope this isn't common enough to come up much in practice.
+ */
+export function workspaceFolderFormatting(ctx: Ctx): Cmd {
+	return async () => {
+		const binaryPath = ctx.lsp.getBinaryPath();
+
+		const workspaceFolder = await selectWorkspaceFolder();
+		if (!workspaceFolder) {
+			return;
+		}
+
+		const allTabsClosed = await closeAllTabs(workspaceFolder);
+		if (!allTabsClosed) {
+			return;
+		}
+
+		const workspaceFolderPath = workspaceFolder.uri.fsPath;
+
+		let stderr = "";
+		let anyErrors = false;
+
+		// i.e., `air format {workspaceFolderPath} --no-color`
+		const args = ["format", workspaceFolderPath, "--no-color"];
+
+		// This should not matter since the path is explicitly supplied, but better to be safe
+		const options = {
+			cwd: workspaceFolderPath,
+		};
+
+		// A promise that resolves when the spawned process closes or errors
+		const finishedFormatting = new Promise<void>((resolve) => {
```

Comment:
Just noting that this should be extracted in a helper if we call a binary again (I'm slightly surprised there is no higher level API for this).

Reply:
I looked and ended up copying what Go does